internal/verifier/clustertime.go

```diff
@@ -0,0 +1,141 @@
+package verifier
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/10gen/migration-verifier/internal/logger"
+	"github.com/10gen/migration-verifier/internal/retry"
+	"github.com/10gen/migration-verifier/internal/util"
+	"github.com/10gen/migration-verifier/mbson"
+	"github.com/pkg/errors"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+	"go.mongodb.org/mongo-driver/mongo"
+)
+
+const opTimeKeyInServerResponse = "operationTime"
+
+// GetClusterTime returns the remote cluster’s cluster time.
+// In so doing it creates a new oplog entry across all shards.
+// The “note” will go into the cluster’s oplog, so keep it
+// short but meaningful.
+func GetClusterTime(
+	ctx context.Context,
+	logger *logger.Logger,
+	client *mongo.Client, // could change to type=any if we need
+) (primitive.Timestamp, error) {
+	retryer := retry.New(retry.DefaultDurationLimit)
+
+	var optime primitive.Timestamp
+
+	// To get the cluster time, we submit a request to append an oplog note
+	// but set an unreasonably-early maxClusterTime. All the shards will fail
+	// the request but still send their current clusterTime to the mongos,
+	// which will return the most recent of those. Thus we can fetch the
+	// most recent cluster time without altering the oplog.
+	err := retryer.RunForTransientErrorsOnly(
+		ctx,
+		logger,
+		func(_ *retry.Info) error {
+			var err error
+			optime, err = runAppendOplogNote(ctx, client, true)
+			if err != nil {
+				err = fmt.Errorf("%w: %w", retry.CustomTransientErr, err)
+			}
+			return err
+		},
+	)
+
+	if err != nil {
+		return primitive.Timestamp{}, err
+	}
+
+	// OPTIMIZATION: We now append an oplog entry--this time for real!--to
+	// cause any lagging shards to output their events.
+	//
+	// Since this is just an optimization, failures here are nonfatal.
+	err = retryer.RunForTransientErrorsOnly(
+		ctx,
+		logger,
+		func(_ *retry.Info) error {
+			var err error
+			optime, err = runAppendOplogNote(ctx, client, false)
+			if err != nil {
+				err = fmt.Errorf("%w: %w", retry.CustomTransientErr, err)
+			}
+			return err
+		},
+	)
+	if err != nil {
+		// This isn't serious enough even to warn on, so leave it at info-level.
+		logger.Info().Err(err).
+			Msg("Failed to append oplog note; change stream may need extra time to finish.")
+	}
+
+	return optime, nil
+}
+
+func runAppendOplogNote(
```

Comment:
I feel like this would be easier to read if it was two different functions: one that's the no-write version (the current doomed=true), and another that's the actutal-write version. I find all the if statements here difficult to follow, plus "doomed" is a weird parameter name (and boolean parameters basically always smell), and if this were `getCurrentTime` and `advanceCurrentTime` or something.

Reply:
Now that you’ve made me look at this again :-P, I find I still don’t fully grok why a 2nd `appendOplogNote` is necessary to prevent a delay from the shards in outputting their events.

I’m just going to simplify it to a single `appendOplogNote`. It seems like that’s all that should be needed anyway, and if I’m wrong, a 10-second wait is miniscule in the grand scheme.

Reply:
Even better! This is way easier to follow now, thanks.